benchmark/utils/general.hpp

```diff
@@ -434,4 +460,180 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * ```
+ * auto timer = get_timer(...);
+ * IterationControl ic(timer);
+ * for(auto status: ic.[warmup_run|run]()){
+ *   timer->start();  // has to be the same timer as passed to ic
+ *   // execute benchmark
+ *   timer->stop();
+ * }
+ * ```
+ * At the beginning of both methods, the timer is reset.
+ * The `status` object exposes the member
+ * - `cur_it`, containing the current iteration number,
+ * and the methods
+ * - `is_finished`, checks if the benchmark is finished,
+ * - `is_last_iteration`, checks if the current iteration is the last one.
+ *
+ */
+class IterationControl {
+    using IndexType = unsigned int;  //!< to be compatible with GFLAGS type
+
+    class run_control;
+
+public:
+    /**
+     * Creates an `IterationControl` object.
+     *
+     * Uses the commandline flags to setup the stopping criteria for the
+     * warmup and timed run.
+     *
+     * @param timer the same timer that is to be used for the timings
+     */
+    explicit IterationControl(std::shared_ptr<Timer> timer) : timer_(timer)
+    {
+        status_warmup_ = {timer, FLAGS_warmup, FLAGS_warmup, 0., 0};
+        if (FLAGS_repetitions == "auto") {
+            status_run_ = {timer, FLAGS_min_repetitions, FLAGS_max_repetitions,
+                           FLAGS_min_runtime};
+        } else {
+            const auto reps =
+                static_cast<unsigned int>(std::stoi(FLAGS_repetitions));
+            status_run_ = {timer, reps, reps, 0., 0};
+        }
+    }
+
+    IterationControl() = default;
+    IterationControl(const IterationControl &) = default;
+    IterationControl(IterationControl &&) = default;
+
+    /**
+     * Creates iterable `run_control` object for the warmup run.
+     *
+     * This run uses always a fixed number of iterations.
+     */
+    run_control warmup_run()
+    {
+        timer_->clear();
+        status_warmup_.cur_it = 0;
+        return run_control{status_warmup_};
+    }
+
+    /**
+     * Creates iterable `run_control` object for the timed run.
+     *
+     * This run may be adaptive, depending on the commandline flags.
+     */
+    run_control run()
+    {
+        timer_->clear();
+        status_run_.cur_it = 0;
+        return run_control{status_run_};
+    }
+
+private:
+    /**
+     * Stores stopping criteria of the adaptive benchmark run as well as the
+     * current iteration number.
+     */
+    struct status {
+        std::shared_ptr<Timer> timer = nullptr;
+
+        IndexType min_it = 0;
+        IndexType max_it = 0;
+        double max_runtime = 0.;
+
+        IndexType cur_it = 0;
+
+        /**
+         * checks if the adaptive run is complete
+         *
+         * the adaptive run is complete if:
+         * - the minimum number of iteration is reached
+         * - and either:
+         *   - the maximum number of repetitions is reached
+         *   - the total runtime is above the threshold
+         * @return completeness state of the adaptive run
```

Comment:
I guess doxygen is fine with this, but we usually have a newline here
nit:
```suggestion
         *   - the total runtime is above the threshold
         *
         * @return completeness state of the adaptive run
```